components/popover_list_members/popover_list_members.tsx

```diff
@@ -0,0 +1,273 @@
+import {localizeMessage} from 'utils/utils.jsx';
+
+import {trackEvent} from 'actions/telemetry_actions.jsx';
+
+import {AddMembersToChanneltreatments} from 'mattermost-redux/constants/config';
+import {Channel} from 'mattermost-redux/types/channels';
+import {UserProfile} from 'mattermost-redux/types/users';
+import {RelationOneToOne} from 'mattermost-redux/types/utilities';
+
+import {ModalData} from 'types/actions';
+
+import ChannelMembersModal from 'components/channel_members_modal';
+import ChannelInviteModal from 'components/channel_invite_modal';
+import OverlayTrigger from 'components/overlay_trigger';
+import Tooltip from 'components/tooltip';
+import Popover from 'components/widgets/popover';
+import PopoverListMembersItem from 'components/popover_list_members/popover_list_members_item';
+
+export type Props = {
+    channel: Channel;
+    statuses: RelationOneToOne<UserProfile, string>;
+    memberCount: number;
+    currentUserId: string;
+    manageMembers: boolean;
+    actions: {
+        openModal: <P>(modalData: ModalData<P>) => void;
+        loadProfilesAndStatusesInChannel: (channelId: string, page?: number, perPage?: number, sort?: string, options?: {active: boolean}) => void;
+        openDirectChannelToUserId: (userId: string) => Promise<{data: Channel}>;
+    };
+    sortedUsers: UserProfile[];
+    addMembersABTest?: string;
+    teamUrl?: string;
+}
+
+export default function PopoverListMembers(props: Props) {
+    const membersListRef: React.RefObject<HTMLDivElement> = React.createRef();
+    const targetRef: React.RefObject<HTMLButtonElement> = React.createRef();
+
+    const [showPopover, setShowPopover] = useState(false);
+
+    const handleShowDirectChannel = (user: UserProfile) => {
+        const {actions} = props;
+        const teammateId = user.id;
+
+        if (teammateId) {
+            actions.openDirectChannelToUserId(teammateId).then(({data}) => {
+                if (data) {
+                    browserHistory.push(props.teamUrl + '/channels/' + data.name);
+                }
+                closePopover();
+            });
+        }
+    };
+
+    const closePopover = () => setShowPopover(false);
+
+    const showMembersModal = (e: React.MouseEvent<HTMLButtonElement>) => {
+        const {actions} = props;
+        e.preventDefault();
+        closePopover();
+
+        const modalData = {
+            modalId: ModalIdentifiers.CHANNEL_MEMBERS,
+            dialogProps: props,
+            dialogType: ChannelMembersModal,
+        };
+        actions.openModal(modalData);
+    };
+
+    const onAddNewMembersButton = (placement: string) => {
+        const {channel, actions} = props;
+        trackEvent('add_members_from_channel_popover', placement);
+
+        const modalData = {
+            modalId: ModalIdentifiers.CHANNEL_INVITE,
+            dialogType: ChannelInviteModal,
+            dialogProps: {channel},
+        };
+        actions.openModal(modalData);
+        closePopover();
+    };
```

Comment:
in this case I would suggest a factory style to make constructing it easier in the components. But this step is totally optional:

```suggestion
    const getOnAddNewMembersButton = (placement: string) => (e: React.MouseEvent<HTMLButtonElement>) => {
        const {channel, actions} = props;

        trackEvent('add_members_from_channel_popover', placement);

        e.preventDefault();
        closePopover();

        const modalData = {
            modalId: ModalIdentifiers.CHANNEL_INVITE,
            dialogType: ChannelInviteModal,
            dialogProps: {channel},
        };

        actions.openModal(modalData);
    };
```

that way you can use it in the component like this:

```
...
onClick={getOnAddNewMembersButton(AddMembersToChanneltreatments.BOTTOM)}
...

Reply:
Makes sense, done